File: native/src/Sprites.swift
// SessionPet — sprite assets (built-ins + drop-in user packs) and rendering.

import AppKit
import Foundation

// MARK: - assets

struct Species {
    let key: String, name: String, emoji: String
    let palette: [String: NSColor], rows: [String]
    // optional hand-authored walk cycle: array of frames, each a rows array
    // (same size as `rows`). Absent → procedural leg-shuffle while walking.
    let walkFrames: [[String]]
}

private func makeSpecies(_ key: String, _ s: [String: Any]) -> Species {
    var pal: [String: NSColor] = [:]
    for (ch, hex) in (s["palette"] as? [String: String]) ?? [:] { pal[ch] = hexColor(hex) }
    let rows = s["rows"] as? [String] ?? []
    let walk = (s["walk"] as? [[String]] ?? []).filter { $0.count == rows.count }
    return Species(key: key, name: s["name"] as? String ?? key,
                   emoji: s["emoji"] as? String ?? "",
                   palette: pal, rows: rows, walkFrames: walk)
}

func loadAssets() -> (order: [String], species: [String: Species]) {
    guard let data = FileManager.default.contents(atPath: assetsPath),
          let root = json(data),
          var order = root["order"] as? [String],
          let dict = root["species"] as? [String: [String: Any]] else {
        fatalError("cannot load \(assetsPath) — run native/export_assets.py")
    }
    var out: [String: Species] = [:]
    for (key, s) in dict { out[key] = makeSpecies(key, s) }
    // user sprite packs: sprites/<key>.json, one species per file, same schema
    // as an assets.json species entry ({"name","emoji","palette","rows"}).
    // Species key = filename stem; packs OVERWRITE built-ins (incl. "egg").
    // Sorted for a deterministic merge order; malformed files are skipped,
    // never fatal.
    let fm = FileManager.default
    for f in ((try? fm.contentsOfDirectory(atPath: spritesDir)) ?? []).sorted()
    where f.hasSuffix(".json") {
        guard let d = fm.contents(atPath: "\(spritesDir)/\(f)"),
              let s = json(d),
              let rows = s["rows"] as? [String], !rows.isEmpty else {
            petLog("sprite pack skipped: \(f)")
            continue
        }
        let key = String(f.dropLast(".json".count))
        out[key] = makeSpecies(key, s)
        // "egg" stays out of the picker — it's the unhatched stage, and picking
        // a sprite is what hatches; an egg.json pack only reskins it
        if key != "egg", !order.contains(key) { order.append(key) }
    }
    return (order, out)
}

let assets = loadAssets()

// MARK: - sprite rendering

func drawSprite(_ key: String, scale: CGFloat, at origin: NSPoint, eyesClosed: Bool,
                mirrored: Bool = false, walkFrame: Int? = nil) {
    guard let sp = assets.species[key] else { return }
    // walk animation: hand-authored frames when the pack ships them,
    // otherwise a procedural 2-frame leg shuffle (bottom rows scissor by
    // one pixel) that works for every species
    var rows = sp.rows
    var shuffle = 0
    if let f = walkFrame {
        if !sp.walkFrames.isEmpty {
            rows = sp.walkFrames[f % sp.walkFrames.count]
        } else {
            shuffle = f % 2 == 0 ? 1 : -1
        }
    }
    let rowCount = rows.count
    let colCount = rows.first?.count ?? 16
    // snap every cell edge to whole pixels: bob/hop offsets are fractional,
    // and unsnapped rects antialias into hairline seams between rows —
    // visible as "stripes" on 1x external monitors. Adjacent cells share the
    // exact same rounded edge, so no gaps and no overlaps by construction.
    var xs = [CGFloat](); xs.reserveCapacity(colCount + 3)
    var ys = [CGFloat](); ys.reserveCapacity(rowCount + 1)
    for i in -1...(colCount + 1) { xs.append((origin.x + CGFloat(i) * scale).rounded()) }
    for j in 0...rowCount { ys.append((origin.y + CGFloat(j) * scale).rounded()) }
    for (y, row) in rows.enumerated() {
        // scissor the feet: last row steps one way, the row above the other
        var stepX = 0
        if shuffle != 0, y >= rowCount - 2 {
            stepX = (y == rowCount - 1 ? shuffle : -shuffle)
        }
        let ry = rowCount - 1 - y  // flip y: pixel row 0 is the TOP
        for (x, ch) in row.enumerated() {
            var c = String(ch)
            if c == "." { continue }
            if eyesClosed && (c == "o" || c == "w") { c = "X" }
            guard let color = sp.palette[c] else { continue }
            color.setFill()
            // mirror x when the pet walks left; +1 offsets into xs, which
            // starts one cell early so the walk shuffle can step past an edge
            let px = (mirrored ? colCount - 1 - x : x) + (mirrored ? -stepX : stepX) + 1
            NSRect(x: xs[px], y: ys[ry],
                   width: xs[px + 1] - xs[px], height: ys[ry + 1] - ys[ry]).fill()
        }
    }
}

func spriteImage(_ key: String, scale: CGFloat) -> NSImage {
    guard let sp = assets.species[key] else { return NSImage() }
    let w = CGFloat(sp.rows.first?.count ?? 16) * scale
    let h = CGFloat(sp.rows.count) * scale
    let img = NSImage(size: NSSize(width: w, height: h))
    img.lockFocus()
    drawSprite(key, scale: scale, at: .zero, eyesClosed: false)
    img.unlockFocus()
    return img
}

File: linux/face/src/sprites.rs
//! Sprite assets (assets.json + drop-in sprites/*.json packs) and cairo
//! rendering. Port of native/src/Sprites.swift.

use gtk::cairo::Context;
use std::collections::HashMap;
use std::fs;

pub struct Species {
    pub name: String,
    pub palette: HashMap<char, (f64, f64, f64)>,
    pub rows: Vec<String>,
    // optional hand-authored walk cycle; absent → procedural leg-shuffle
    pub walk_frames: Vec<Vec<String>>,
}

pub struct Assets {
    pub order: Vec<String>,
    pub species: HashMap<String, Species>,
}

fn hex_color(hex: &str) -> (f64, f64, f64) {
    let h = hex.trim_start_matches('#');
    if h.len() != 6 {
        return (1.0, 1.0, 1.0);
    }
    match u32::from_str_radix(h, 16) {
        Ok(v) => (
            ((v >> 16) & 0xFF) as f64 / 255.0,
            ((v >> 8) & 0xFF) as f64 / 255.0,
            (v & 0xFF) as f64 / 255.0,
        ),
        Err(_) => (1.0, 1.0, 1.0),
    }
}

fn make_species(key: &str, s: &serde_json::Value) -> Species {
    let mut palette = HashMap::new();
    if let Some(pal) = s.get("palette").and_then(|v| v.as_object()) {
        for (ch, hex) in pal {
            if let (Some(c), Some(h)) = (ch.chars().next(), hex.as_str()) {
                palette.insert(c, hex_color(h));
            }
        }
    }
    let rows: Vec<String> = s
        .get("rows")
        .and_then(|v| v.as_array())
        .map(|a| a.iter().filter_map(|r| r.as_str().map(String::from)).collect())
        .unwrap_or_default();
    let walk_frames: Vec<Vec<String>> = s
        .get("walk")
        .and_then(|v| v.as_array())
        .map(|frames| {
            frames
                .iter()
                .filter_map(|f| {
                    let fr: Vec<String> = f
                        .as_array()?
                        .iter()
                        .filter_map(|r| r.as_str().map(String::from))
                        .collect();
                    (fr.len() == rows.len()).then_some(fr)
                })
                .collect()
        })
        .unwrap_or_default();
    Species {
        name: s
            .get("name")
            .and_then(|v| v.as_str())
            .unwrap_or(key)
            .to_string(),
        palette,
        rows,
        walk_frames,
    }
}

/// Built-ins from assets.json (embedded or repo copy), then user packs from
/// sprites/<key>.json — packs OVERWRITE built-ins; malformed files skipped.
pub fn load_assets(assets_json: &str, sprites_dir: Option<&std::path::Path>) -> Assets {
    let root: serde_json::Value =
        serde_json::from_str(assets_json).expect("assets.json is malformed");
    let mut order: Vec<String> = root
        .get("order")
        .and_then(|v| v.as_array())
        .map(|a| a.iter().filter_map(|s| s.as_str().map(String::from)).collect())
        .unwrap_or_default();
    let mut species = HashMap::new();
    if let Some(dict) = root.get("species").and_then(|v| v.as_object()) {
        for (key, s) in dict {
            species.insert(key.clone(), make_species(key, s));
        }
    }
    if let Some(dir) = sprites_dir {
        let mut files: Vec<_> = fs::read_dir(dir)
            .map(|rd| rd.filter_map(|e| e.ok().map(|e| e.path())).collect())
            .unwrap_or_default();
        files.sort();
        for f in files {
            if f.extension().and_then(|e| e.to_str()) != Some("json") {
                continue;
            }
            let Ok(text) = fs::read_to_string(&f) else { continue };
            let Ok(s) = serde_json::from_str::<serde_json::Value>(&text) else {
                continue;
            };
            let sp_rows = s.get("rows").and_then(|v| v.as_array());
            if sp_rows.map_or(true, |r| r.is_empty()) {
                continue;
            }
            let key = f
                .file_stem()
                .and_then(|s| s.to_str())
                .unwrap_or_default()
                .to_string();
            species.insert(key.clone(), make_species(&key, &s));
            // "egg" stays out of the picker — picking a sprite is what hatches
            if key != "egg" && !order.contains(&key) {
                order.push(key);
            }
        }
    }
    Assets { order, species }
}

/// Draw at (ox, oy_top) — cairo y grows DOWN, row 0 is the sprite's top row.
#[allow(clippy::too_many_arguments)]
pub fn draw_sprite(
    ctx: &Context,
    sp: &Species,
    scale: f64,
    ox: f64,
    oy_top: f64,
    eyes_closed: bool,
    mirrored: bool,
    walk_frame: Option<usize>,
) {
    let mut rows = &sp.rows;
    let mut shuffle: i64 = 0;
    if let Some(f) = walk_frame {
        if !sp.walk_frames.is_empty() {
            rows = &sp.walk_frames[f % sp.walk_frames.len()];
        } else {
            // procedural 2-frame leg shuffle: bottom rows scissor by one px
            shuffle = if f % 2 == 0 { 1 } else { -1 };
        }
    }
    let row_count = rows.len() as i64;
    let col_count = rows.first().map_or(16, |r| r.chars().count() as i64);
    // snap cell edges to whole pixels — unsnapped rects antialias into
    // hairline seams between rows (same fix as the Mac renderer)
    let xs: Vec<f64> = (-1..=col_count + 1)
        .map(|i| (ox + i as f64 * scale).round())
        .collect();
    let ys: Vec<f64> = (0..=row_count)
        .map(|j| (oy_top + j as f64 * scale).round())
        .collect();
    for (y, row) in rows.iter().enumerate() {
        let y = y as i64;
        let mut step_x: i64 = 0;
        if shuffle != 0 && y >= row_count - 2 {
            step_x = if y == row_count - 1 { shuffle } else { -shuffle };
        }
        for (x, ch) in row.chars().enumerate() {
            let x = x as i64;
            let mut c = ch;
            if c == '.' {
                continue;
            }
            if eyes_closed && (c == 'o' || c == 'w') {
                c = 'X';
            }
            let Some(&(r, g, b)) = sp.palette.get(&c) else { continue };
            ctx.set_source_rgb(r, g, b);
            let px = (if mirrored { col_count - 1 - x } else { x }
                + if mirrored { -step_x } else { step_x }
                + 1) as usize;
            let yy = y as usize;
            ctx.rectangle(
                xs[px],
                ys[yy],
                xs[px + 1] - xs[px],
                ys[yy + 1] - ys[yy],
            );
            let _ = ctx.fill();
        }
    }
}
